[User History] Track bulk upload changes

mkangia opened the pull request:
## Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->

To reviewer,
It would be good to take quick a look at the final state of code first, mainly [user_importer](https://github.com/dimagi/commcare-hq/pull/29935/files#diff-3875199418c65c97fec9e657efd64668638b4bb447d3a24a1ad3dd42e14064c4) to get a sense of how eventually code came out to be and then start commit wise. There is quite some refactor and a couple of corrections in implementation which would be easy to follow along in commits.
Thanks

This PR, Builds on #29887 to add tracking for updates done via bulk user upload.
- adds `UserChangeLogger` to log
changes to user properties with the new value
change messages for updates like password reset
useful info messages for names of roles and locations user is updated to have
- [refactor] add `CommCareUserImporter` and `WebUserImporter` to extract out import for each user and handle logging

Refactor here that has been done in the end to make things cleaner, after the tests were added.
The refactor can definitely go much further, but stopped at this point for now.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
This does not change the functionality of user bulk upload at all but just adds logging of user changes. Though there is code refactor. 

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
`corehq.apps.user_importer.tests.test_importer` have really extensive tests around user importer and I have added the expectations from logging perspective on tests to cover possible scenarios.

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
https://dimagi-dev.atlassian.net/browse/QA-3142

### Safety story
<!--
Describe any other pieces to the safety story including
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->
Tests are good. Also the refactor is mainly moving the code and renaming, done commit wise for easier follow up as small changes.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

same as base PR

mkangia committed:
better name

mkangia committed:
track user history for commcare users

- add helper class
- remove redundant log_user_role_update
- track create as well via helper class
- allow update log with only messages as well

mkangia committed:
better naming

mkangia committed:
remove unused flag

mkangia committed:
log updates for mobile user's web user

mkangia committed:
save only when necessary

mkangia committed:
log changes for web users

snopoke commented on corehq/apps/user_importer/helpers.py:
```diff
@@ -9,3 +12,68 @@ def spec_value_to_boolean_or_none(user_spec_dict, key):
+
+    def save(self):
+        self._include_user_data_changes()
+        if self.is_new_user:
+            log_user_change(
+                self.domain,
+                self.user,
+                changed_by_user=self.changed_by_user,
+                changed_via=self.changed_via,
+                message=". ".join(self.messages),
+                action=UserModelAction.CREATE,
+            )
+        else:
+            if not self._save:
+                return
+            log_user_change(
+                self.domain,
+                self.user,
+                changed_by_user=self.changed_by_user,
+                changed_via=self.changed_via,
+                message=". ".join(self.messages),
+                fields_changed=self.fields_changed
+            )
```
Could be simplified:
```suggestion
        if self.is_new_user or self._save:
            action = UserModelAction.CREATE if self.is_new_user else UserModelAction.UPDATE
            fields_changed = None if self.is_new_user else self.fields_changed
            log_user_change(
                self.domain,
                self.user,
                changed_by_user=self.changed_by_user,
                changed_via=self.changed_via,
                message=". ".join(self.messages),
                fields_changed=fields_changed,
                action=action,
            )
```

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/29935/commits/dc98af9ada3e684523c90bbca5e2d0fcccf098cf

snopoke commented on corehq/apps/user_importer/importer.py:
```diff
@@ -527,60 +545,108 @@ def create_or_update_users_and_groups(upload_domain, user_specs, upload_user, gr
-
+                        user_change_logger.add_changes({'assigned_location_ids': location_ids})
+                        if location_ids:
+                            location_names = list(SQLLocation.active_objects.filter(
```
can you not get this from `domain_info.location_cache` instead of doing another lookup?

snopoke replied:
Also wondering if this couldn't be moved to the logger to make the logging less 'invasive' on the import code:
```
user_change_logger.add_location_change_info(location_ids, domain_info.location_cache)
```

mkangia replied:
thank you, totally missed about the caching importer has handy. Will incorporate that.

snopoke commented on corehq/apps/user_importer/importer.py:
```diff
@@ -527,60 +545,108 @@ def create_or_update_users_and_groups(upload_domain, user_specs, upload_user, gr
+                    if new_role:
+                        user_change_logger.add_info(_(f"Role: {new_role.name}[{new_role.get_id}]"))
+                    else:
+                        user_change_logger.add_change_message("Role: None")
```
similar to above - this logic could be moved to the logger

snopoke commented on corehq/apps/user_importer/importer.py:
```diff
@@ -771,10 +858,13 @@ def remove_invited_web_user(domain, username):
 
 
-def remove_web_user_from_domain(domain, user, username, upload_user, is_web_upload=False):
+def remove_web_user_from_domain(domain, user, username, upload_user, user_change_logger=None,
```
when would `user_change_logger` be None?

mkangia replied:
`remove_web_user_from_domain` can be None when there is no corresponding web user for a mobile worker.
https://github.com/dimagi/commcare-hq/blob/1779b9e431b184206472311a326d2b02732c76fd/corehq/apps/user_importer/importer.py#L602

Similarly for other places where logger can be None.

The importer logic shuffles between Invitation and users for taking necessary actions. And the logging is only for users and not for invitations, hence the unfortunate branching with logger being `None`.

snopoke commented on corehq/apps/user_importer/importer.py:
```diff
@@ -311,6 +315,8 @@ def create_or_update_web_user_invite(email, domain, role_qualified_id, upload_us
     )
     if invite_created and send_email:
         invite.send_activation_email()
+    if invite_created and user_change_logger:
```
when would `user_change_logger` be None?

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/29935/commits/e8736f5060bb376fd3427682e5f662ea7672f8a7#r656096471

mkangia committed:
[refactor] extract setup_commcare_user_for_import

mkangia committed:
[refactor] add CommCareUserImporter

mkangia committed:
[refactor] rename

web_user -> web_user_username
current_user -> web_user

mkangia committed:
[refactor] add WebUserImporter

mkangia committed:
fix message

mkangia committed:
remove unused imports

mkangia committed:
get locations from cache

mkangia committed:
simplify

mkangia committed:
keep user save out of importer to be same as WebUserImporter

mkangia committed:
make WebUserImporter simpler

mkangia committed:
add useful documentation

mkangia committed:
check for number present before update

mkangia commented on corehq/apps/user_importer/importer.py:
```diff
@@ -484,13 +485,23 @@ def create_or_update_commcare_users_and_groups(upload_domain, user_specs, upload
+                    # always call this to set phone number as default if needed
+                    user.add_phone_number(fmt_phone_number, default=True)
+                    if fmt_phone_number not in user.phone_numbers:
+                        user_change_logger.add_change_message(_(f"Added phone number {fmt_phone_number}"))
```
Corrected later to check if number is already present before updating
https://github.com/dimagi/commcare-hq/pull/29935/commits/632e442311538caa417cbbd01f19d8486d2231bb

mkangia committed:
delay loggin primary location changes

mkangia commented on corehq/apps/user_importer/importer.py:
```diff
@@ -774,24 +780,32 @@ def create_or_update_web_users(upload_domain, user_specs, upload_user, update_pr
             current_user.unset_location(domain, commit=False)
+            user_change_logger.add_changes({'location_id': current_user.location_id})
+            if current_user.location_id:
+                user_change_logger.add_info(_(f"Primary location: {current_user.get_sql_location(domain).name}"))
```
this was later moved to be after the `if locations_updated:` block because the primary location can get re-set in that section.
https://github.com/dimagi/commcare-hq/pull/29935/commits/ebad364ef0537a2a96220efe60a770072d372b6c

mkangia committed:
strengthen tests

mkangia committed:
update WebUser locations updates

- use get_sql_location to get the correct primary location
- save location updates like roles since web user has locations on membership

mkangia committed:
use available helper method

mkangia commented:
Pinging @orangejenny @stephherbers for inputs, if any from their recent work with user uploads.

orangejenny reviewed (commented):
Just got through the first few commits, will finish reviewing next week.

calellowitz commented:
How does this interact with the `UserUploadRecord`? It seems like it tracks a lot of the same information, but perhaps in an easier format. Do we still need both?

mkangia commented on corehq/apps/user_importer/importer.py:
```diff
@@ -484,13 +485,23 @@ def create_or_update_commcare_users_and_groups(upload_domain, user_specs, upload
+                    # always call this to set phone number as default if needed
+                    user.add_phone_number(fmt_phone_number, default=True)
+                    if fmt_phone_number not in user.phone_numbers:
+                        user_change_logger.add_change_message(_(f"Added phone number {fmt_phone_number}"))
```
oh, I thought f strings worked with translations. Will take a look at updating this.

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/29935/commits/f3950d7abea9aa7e833b531e35e2d30936d195ae

mkangia commented on corehq/apps/user_importer/helpers.py:
```diff
@@ -9,3 +12,60 @@ def spec_value_to_boolean_or_none(user_spec_dict, key):
+        self.messages = []
+
+    def add_changes(self, changes):
+        # ignore for new user since the whole user doc is logged for a new user
```
Yes, you are right. It should not be needed.

Would like to hear thoughts on this.
Since the copy of the doc is kept in `init`, and it isn't a costly operation, how about keeping it simple, without the `if` block and then leave the methods to decide if they need it, leaving the awareness of distinction in operations for new/existing objects, only where needed.

snopoke replied:
It may not be that expensive as a single operation but since this is part of an import it needs to be done for each user being updated and the cost could add up. I'd be pro not setting this prop (or setting it to None) for new users.

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/29935/commits/07746469f1d10d6907dee5ca1d547c6f3eb71cb0

mkangia commented:
Hey @calellowitz 

> How does this interact with the UserUploadRecord? It seems like it tracks a lot of the same information, but perhaps in an easier format. Do we still need both?

UserUploadRecord is still needed since this does not store what actually was uploaded but only what was changed. So its a subset of that. Raw information of upload could still be useful in answering queries about upload.

snopoke commented on corehq/apps/user_importer/helpers.py:
```diff
@@ -9,3 +16,283 @@ def spec_value_to_boolean_or_none(user_spec_dict, key):
+        self.messages.append(info)
+        self._save = True
+
+    def save(self):
```
nit: how about changing this to `save_log` since this isn't a model itself.

mkangia replied:
good point. This was named originally as save because it was handling user save and then log save if needed, but then because of multiple user.save calls outside of web importer, for web user uploads, it didn't make sense for user.save to be here, though the method name persisted.

Considering that state now, I am thinking this would be better as a context manager so that the logs are automatically saved on exit.
```
with CommCareUserImporter() as importer:
    # update first name
    # update role
    # etc etc
```
Renamed to `save_log` for now.

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/29935/commits/c7e526f7f23045a00126aa4456157fbf80ac45c9

snopoke replied:
It doesn't look like this function name got changed, only the ones in the importer classes.

snopoke commented on corehq/apps/user_importer/helpers.py:
```diff
@@ -9,3 +16,283 @@ def spec_value_to_boolean_or_none(user_spec_dict, key):
+            else:
+                self.logger.add_change_message("Role: None")
+
+        # ToDo: save log before saving user
```
is this todo still valid?

mkangia replied:
I kept a note for me to reconsider this. If the user is saved, but logs are not, we would not have the history. Though the reverse is harmful as well.
Any thoughts about that?
Is this okay as it is, or better to put this in a transaction?

snopoke replied:
Couch doesn't have transactions. I think the best option is to save the user first and then the log. If the log fails it should report to Sentry.

mkangia replied:
Cool,

removed ToDo https://github.com/dimagi/commcare-hq/pull/29935/commits/091fa00a9422f1837f4a9e535f1f2d5c2ca25d0c

snopoke commented on corehq/apps/user_importer/helpers.py:
```diff
@@ -9,3 +16,283 @@ def spec_value_to_boolean_or_none(user_spec_dict, key):
+        if self.role_updated:
+            self.user.set_role(self.user_domain, role_qualified_id)
+
+    def save(self):
```
nit: `save` doesn't seem like the right name here. I think this could be `save_log` as well.

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/29935/commits/c7e526f7f23045a00126aa4456157fbf80ac45c9

snopoke commented on corehq/apps/user_importer/importer.py:
```diff
@@ -457,131 +463,89 @@ def create_or_update_users_and_groups(upload_domain, user_specs, upload_user, gr
-
-                if web_user:
-                    user.update_metadata({'login_as_user': web_user})
+                    commcare_user_importer.update_role(role, domain_info)
```
why not pass in `role_qualified_id` here so that this matches the `WebUserImporter.update_role` implementation. I think would then have exactly the same implementation and could be put in a super class maybe?

mkangia replied:
thank you. role_qualified_id was already available, so we can just pass that itself

https://github.com/dimagi/commcare-hq/pull/29935/commits/2318dd38c185b16a60d95cfd3a05acaa91e3ddb1

and
https://github.com/dimagi/commcare-hq/pull/29935/commits/d4c56fd788915b15bef29ba535a4e90eb9620144

snopoke commented on corehq/apps/user_importer/helpers.py:
```diff
@@ -9,3 +16,283 @@ def spec_value_to_boolean_or_none(user_spec_dict, key):
+                _(f"Primary location: {location_info}"))
+
+    def save(self):
+        # Tracking for role is done post user save to have role setup correctly on save
+        if self.role_updated:
+            new_role = self.user.get_role(domain=self.user_domain)
+            if new_role:
+                self.logger.add_info(_(f"Role: {new_role.name}[{new_role.get_id}]"))
+            else:
+                self.logger.add_change_message("Role: None")
+        self.logger.save()
```
Seems identical to `CommCareUserImporter.save` - maybe put it in a base class?

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/29935/commits/a884b422896976e6797e704d551dfc27ca3fb5c3

snopoke commented on corehq/apps/user_importer/helpers.py:
```diff
@@ -9,3 +16,283 @@ def spec_value_to_boolean_or_none(user_spec_dict, key):
+            if current_primary_location_id != location_id:
+                users_updated_primary_location_name = get_user_primary_location_name(self.user, self.user_domain)
+                self.logger.add_info(
+                    _(f"Primary location: {users_updated_primary_location_name}[{location_id}]"))
```
Translation doesn't work with f-strings.

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/29935/commits/f3950d7abea9aa7e833b531e35e2d30936d195ae

snopoke reviewed (commented):
Generally looks fine. Only small comments.

mkangia committed:
better name

mkangia committed:
pass qualified id already available

mkangia committed:
extract BaseUserImporter

- remove redundant is_web_users_upload

mkangia committed:
move to common and fix documentation

mkangia commented:
Thank you for the review @snopoke , responded or addressed to feedback.

orangejenny committed:
Merge branch 'master' into mk/user-history/bulk-upload

mkangia committed:
merge mk/user-history/base; resolve conflict in users/test_importer

mkangia committed:
fix translations

mkangia commented on corehq/apps/user_importer/helpers.py:
```diff
@@ -9,3 +12,60 @@ def spec_value_to_boolean_or_none(user_spec_dict, key):
+                fields_changed=self.fields_changed
+            )
+
+    def _include_user_data_changes(self):
```
It looks that would already be covered, since
1. profile is updated as a part of user_data
2. and we track any changes to user_data [here](https://github.com/dimagi/commcare-hq/blob/f3950d7abea9aa7e833b531e35e2d30936d195ae/corehq/apps/user_importer/helpers.py#L63)
Though i see now that since that is happening in logger and not in importer because the original doc is available with the logger, it's hidden, which is not great. Going to see and move it to the importer instead

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/29935/commits/55958b87cbd5df577474352d4bbae32497794f7f

orangejenny replied:
The existing `user_data` handling will only show the profile id, won't it? It'd be helpful to be able to see the name.

mkangia replied:
Thank you. 

https://github.com/dimagi/commcare-hq/pull/29935/commits/983593f86efc0c926bfe44e716365ec448462242

It was a bit confusing that the implementation would look for profile name under "commcare_profile" column but one could still update user data by passing an ID for profile and that is what tests are using. 
So i tried to cover both cases for logging.

orangejenny replied:
Sounds like the original tests aren't realistic, which is probably my bad.

mkangia committed:
include user data changes in importer and not logger

mkangia committed:
Merge branch 'mk/user-history/base' into mk/user-history/bulk-upload

snopoke commented on corehq/apps/user_importer/helpers.py:
```diff
@@ -9,3 +16,269 @@ def spec_value_to_boolean_or_none(user_spec_dict, key):
+        self.logger.save()
+
+    def _include_user_data_changes(self):
+        # ToDo: consider putting just the diff
```
either you should do this or remove the TODO (unless you plan to do it in a follow up PR?)

mkangia replied:
I have not been able to find a good way to diff for this, so i would like to keep this note and take it in a follow up PR, in case I am not able to get to it before merging the base branch.
thanks for checking on this.

snopoke commented on corehq/apps/user_importer/helpers.py:
```diff
@@ -9,3 +16,269 @@ def spec_value_to_boolean_or_none(user_spec_dict, key):
+class WebUserImporter(BaseUserImporter):
+    def add_to_domain(self, role_qualified_id, location_id):
+        self.user.add_as_web_user(self.user_domain, role=role_qualified_id, location_id=location_id)
+        self.logger.add_change_message(_("Added as web user"))
```
This could use more detail: `"Added user to domain '{domain}' with role '{role}' and location '{location}'"`

mkangia replied:
I corrected the documentation added. Role and location are assigned only if web user is already a member of the domain, so that would not happen here.

Also generally, Role and location information will get added when and if those are assigned.

Just thought,
My sense with these messages is that it should avoid implementation context and should be specific to what changed. So not saying that `User was added with foo role and bar location`, but saying, `User was added. Role foo. Location Bar` to have a simpler and common way of adding them from all places.

mkangia replied:
Documentation updated https://github.com/dimagi/commcare-hq/pull/29935/commits/a6dabaccbe0b14804b550c86b296b3d6ceea7957

snopoke replied:
I don't quite follow, if the user is already a member of the domain and the role and location are handled elsewhere they why are we calling this method which does all 3 of those things?

My point is that "Added as web user" is not going to mean anything to a user reading the logs and it would be better to give a message that a user is going to understand.

mkangia replied:
> I don't quite follow, if the user is already a member of the domain and the role and location are handled elsewhere they why are we calling this method which does all 3 of those things?

Ah, I see. You are right. We are adding user along with a role and possibly with a location. So we should also track that here.

https://github.com/dimagi/commcare-hq/pull/29935/commits/29d0abf4f7291a23e10dd012c46f078a395245d7

> My point is that "Added as web user" is not going to mean anything to a user reading the logs and it would be better to give a message that a user is going to understand.

Just generally, its referring to that the user was added as a member to the domain. If we were to find out when a user started getting access to a domain, this would be the log to look at. Does that make sense?

mkangia replied:
also, as a follow up https://github.com/dimagi/commcare-hq/pull/29935/commits/9fa95b0ab6b284a953702277d51520d248b994ff

snopoke replied:
Yes it does, I'm just thinking about how this will show up in the report. Maybe we could re-word it to "Added as web user to this domain"? Assuming that the context of the report is the domain that the user was added to?

I'm also thinking that if you were just looking at the logs for the web user (across all domains) you'd need to consider which domain the log was created for in order to know which domain they were added to.

mkangia replied:
>  Maybe we could re-word it to "Added as web user to this domain"? Assuming that the context of the report is the domain that the user was added to?

Adding domain name in this itself will be useful when we have multi domain report. Users for a master domain, looking at report for users in all linked domains. 

For example:
When filtering on a master domain,
User should see changes done for users in this domain and the linked domains,
by users belonging to this domain and the linked domains,
Where logs would look like,
User    |              Message
User1  | Added as web user to cape-town. Role: SSE
User1  | Added as web user to boston. Role: Read-Only. Primary Location: DimagiOffice

> I'm also thinking that if you were just looking at the logs for the web user (across all domains) you'd need to consider which domain the log was created for in order to know which domain they were added to.

I see your point. Similar to "by_domain", we should also have a "for_domain". 
I was thinking but don't see the use case yet. When anything is updated on the user that is domain specific, like location, role, they all have a "domain" attribute on it, I believe, mostly through domain membership. Can't totally rule out the need for adding "for_domain" in future.

snopoke commented on corehq/apps/user_importer/importer.py:
```diff
@@ -771,15 +755,20 @@ def remove_invited_web_user(domain, username):
         user.delete_domain_membership(domain)
         user.save()
+        if user_change_logger:
+            user_change_logger.add_change_message(_("Removed from domain"))
```
```suggestion
            user_change_logger.add_change_message(_("Removed from domain '{domain}'").format(domain=domain))
```

mkangia replied:
Thanks for pointing that out. 
Updated this and other places that could use the domain name
https://github.com/dimagi/commcare-hq/pull/29935/commits/7be66111d94d17a023e2161bb84ec416d876e7c0

orangejenny commented on corehq/apps/user_importer/importer.py:
```diff
@@ -687,6 +668,31 @@ def create_or_update_commcare_users_and_groups(upload_domain, user_specs, upload
     return ret
 
 
+def _setup_commcare_user_for_import(domain, user_id, username, is_account_confirmed, web_user, password,
```
How about "get_or_create" to be more specific than "setup"?

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/29935/commits/44c4898e0c54d7786ae50ca44381a9f6a8ab2ecf

a catch to that is that the user is actually not "created" because the record is not saved yet. Was thinking of `get_or_init` though that would be confusing as well.

orangejenny replied:
That's a good point. `get_or_init` sounds fine, too.

orangejenny commented on corehq/apps/user_importer/importer.py:
```diff
@@ -448,7 +448,7 @@ def create_or_update_commcare_users_and_groups(upload_domain, user_specs, upload
             role = row.get('role', None)
             profile = row.get('user_profile', None)
-            web_user = row.get('web_user')
+            web_user_username = row.get('web_user')
```
yes!

orangejenny commented on corehq/apps/user_importer/importer.py:
```diff
@@ -398,6 +398,15 @@ def format_location_codes(location_codes):
                                                update_progress=None):
+    """"
+    Creates and Updates CommCare Users
+    For the associated web user username passed, for each CommCareUser
```
Nitpick: the associated web user logic is a weird USH-only workflow, it's a little odd for it to be part of the primary function description.

mkangia replied:
Not sure if it should not be present in the documentation because it's specific to a project unless we can extract all that logic to a method and put this part there, though i would still prefer this being here so that its clear that one should think about this workflow if they are making changes to this method.

Surprised, that this is not behind a feature flag.

orangejenny replied:
Fair.

It's a "secret" feature, since to access it you have to know about a magical user data key.

orangejenny commented on corehq/apps/user_importer/helpers.py:
```diff
@@ -9,3 +16,269 @@ def spec_value_to_boolean_or_none(user_spec_dict, key):
+        )
+
+        location_ids = find_location_id(location_codes, domain_info.location_cache)
+        users_current_primary_location_id = self.user.location_id
```
nitpick: `user_current_primary_location_id` instead of `users_current_primary_location_id`

mkangia replied:
thank you. 
Updated this and one more
https://github.com/dimagi/commcare-hq/pull/29935/commits/d3898ed2ae37199cdacd96bbafc405d7621dc4e3

mkangia committed:
skip original doc for new users

mkangia committed:
add profile name as info

mkangia committed:
add domain name

mkangia committed:
fix documentation

mkangia committed:
remove redundant todo

mkangia committed:
users -> user

mkangia committed:
better name

mkangia commented:
@snopoke @orangejenny 
Thank you for the review, I have tried to address all concerns.
requesting a re-review.

mkangia committed:
Merge branch 'mk/user-history/base' into mk/user-history/bulk-upload

mkangia committed:
track role and location change; fix documentation

mkangia committed:
dedup

orangejenny commented on corehq/apps/user_importer/helpers.py:
```diff
@@ -32,7 +32,11 @@ def __init__(self, domain, user, is_new_user, changed_by_user, changed_via):
-        self.original_user_doc = self.user.to_json()
+        if not is_new_user:
+            self.original_user_doc = self.user.to_json()
+        else:
```
Isn't this branch unnecessary?

mkangia replied:
https://github.com/dimagi/commcare-hq/commit/07746469f1d10d6907dee5ca1d547c6f3eb71cb0#diff-2f11ee2db6cd89244e3a42d3ed2a050b2bb42fde7eb86234a123855a6d75bcc2R120

Ya.

orangejenny commented:
@mkangia Thank you for the quick followup, all of my feedback is addressed.

mkangia committed:
Merge branch 'mk/user-history/base' into mk/user-history/bulk-upload

mkangia commented:
Hey @snopoke 
Can you take another pass on review please?

snopoke commented on corehq/apps/user_importer/helpers.py:
```diff
@@ -159,6 +164,11 @@ def update_user_data(self, data, uncategorized_data, profile, domain_info):
                 self.user.pop_metadata(key)
 
+        if self.user.user_data.get(PROFILE_SLUG) and self.user.user_data[PROFILE_SLUG] != current_profile_id:
+            self.logger.add_info(_("CommCare Profile: {profile_name}").format(
```
what's the guideline for when to use `add_info` vs `add_change_message`?

mkangia replied:
I should document this or better to standardize this with better methods in class or documentation there, my approach has been
1. `add_changes`: changes to actual user properties, for audit, ignored for new user. Example: `first_name`
2. `add_change_message`: change messages to users that are not exactly properties, ignored for new user. Example: "Password reset" / "Added phone number 9999999999"
3. `add_info`: any useful information, mainly for display, irrespective of new/old user. Example: "Role: Read Only[role-id]"

snopoke commented on corehq/apps/user_importer/importer.py:
```diff
@@ -306,7 +306,7 @@ def create_or_update_web_user_invite(email, domain, role_qualified_id, upload_us
         invite.send_activation_email()
     if invite_created and user_change_logger:
-        user_change_logger.add_change_message(_("Invited to domain"))
+        user_change_logger.add_change_message(_("Invited to domain {domain_name}").format(domain_name=domain))
```
nit: quote domain name
```suggestion
        user_change_logger.add_change_message(_("Invited to domain '{domain_name}'").format(domain_name=domain))
```

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/29935/commits/15e180758f65e1ac9a952e36fc8de8bedf73fa46

snopoke commented on corehq/apps/user_importer/helpers.py:
```diff
@@ -215,19 +225,28 @@ def _fmt_phone(phone_number):
-        self.logger.add_change_message(_("Added as web user"))
+        self.role_updated = bool(role_qualified_id)
+
+        self.logger.add_change_message(_("Added as web user to {domain_name}").format(
```
nit: quote domain name
```suggestion
        self.logger.add_change_message(_("Added as web user to '{domain_name}'").format(
```

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/29935/commits/15e180758f65e1ac9a952e36fc8de8bedf73fa46

snopoke commented on corehq/apps/user_importer/importer.py:
```diff
@@ -761,7 +771,9 @@ def remove_web_user_from_domain(domain, user, username, upload_user, user_change
             remove_invited_web_user(domain, username)
             if user_change_logger:
-                user_change_logger.add_change_message(_("Invitation revoked"))
+                user_change_logger.add_change_message(_("Invitation revoked from {domain_name}").format(
```
```suggestion
                user_change_logger.add_change_message(_("Invitation revoked from '{domain_name}'").format(
```

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/29935/commits/15e180758f65e1ac9a952e36fc8de8bedf73fa46

snopoke commented on corehq/apps/user_importer/importer.py:
```diff
@@ -771,4 +783,6 @@ def remove_web_user_from_domain(domain, user, username, upload_user, user_change
         user.save()
         if user_change_logger:
-            user_change_logger.add_change_message(_("Removed from domain"))
+            user_change_logger.add_change_message(_("Removed from domain {domain_name}").format(
```
```suggestion
            user_change_logger.add_change_message(_("Removed from domain '{domain_name}'").format(
```

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/29935/commits/15e180758f65e1ac9a952e36fc8de8bedf73fa46

snopoke reviewed (approved):
small nits / comments but otherwise GTG

mkangia committed:
quote domain names and be consistent with message structure

mkangia committed:
add subscription removed in other PR

https://github.com/dimagi/commcare-hq/pull/29989/files#diff-228026314a99cb3fbdde4cae7c03b0ea76e8732b93533407a9856494075319bfL40

mkangia commented:
Thanks for the review @snopoke 

Did a single [commit](https://github.com/dimagi/commcare-hq/pull/29935/commits/15e180758f65e1ac9a952e36fc8de8bedf73fa46) for suggestions on quoting domain name.
